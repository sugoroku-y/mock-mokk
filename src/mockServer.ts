import express from 'express';

export function mockServer(
  port: number,
  index: string[],
  files: [string, string][]
) {
  // serverの設定
  const app = express();

  const staticOptions: Parameters<typeof express.static>[1] = {index};
  for (const [location, path] of files) {
    app.use(location, express.static(path, staticOptions));
  }
  // URLで指定されたファイルが見つからない場合用のエントリー
  app.use((req, res) => {
    console.error(`ファイルが見つかりません。: ${req.path}`);
    res.sendStatus(404);
  });

  // サーバー起動
  app.listen(port, () => {
    console.log(`${port}番ポートで開始します。`);
  });
  return app;
}
