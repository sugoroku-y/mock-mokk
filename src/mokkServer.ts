import express from 'express';
import { Server } from 'http';

export function mokkServer(
  port: number,
  index: string[],
  files: [string, string][]
): Promise<Server> {
  // serverの設定
  const app = express();

  const staticOptions = { index };
  for (const [location, path] of files) {
    app.use(location, express.static(path, staticOptions));
  }

  // URLで指定されたファイルが見つからない場合用のエントリー
  app.use((req, res) => {
    console.error(`ファイルが見つかりません。: ${req.path}`);
    res.sendStatus(404);
  });

  // サーバー起動
  return new Promise<Server>((resolve, reject) => {
    const server = app
      .listen(port, () => {
        console.log(`${port}番ポートで開始します。`);
        resolve(server);
      })
      .on('error', err => reject(err));
  });
}
