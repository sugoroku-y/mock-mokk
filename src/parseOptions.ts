import { existsSync } from 'fs';
import { homedir } from 'os';
import { resolve, sep } from 'path';
import { parse, helpString, unnamed } from 'optionalist';

declare global {
  interface String {
    replace(
      re: RegExp,
      replacer: (match: string, ...captures: (string | undefined)[]) => string,
    ): string;
  }
}

/**
 * パスを指定する文字列を絶対パスに解決する。
 *
 * 文字列中に以下のパターンがあればそれぞれ置換します。
 *
 * - 先頭にある`$/`はホームディレクトリに
 * - `$ENV`もしくは`${ENV}`は環境変数ENVの値に
 * - `$$`は文字`$`そのものに
 *
 * @param {string} path
 * @returns {string}
 */
function resolvePathSpec(path: string): string {
  return resolve(
    path.replace(
      /^\$\/|\$\$|\$\{(\w+)}|\$(\w+)/g,
      (match, bracketedName, bareName) => {
        if (match === '$/') {
          // 先頭から`$/`で始まっていたらホームディレクトリ
          return homedir() + sep;
        }
        if (match === '$$') {
          // `$$`は`$`1文字
          return '$';
        }
        const name = bracketedName ?? bareName;
        // `$ENV`、`${ENV}`は環境変数ENVの値に、ENVがなければ空文字列
        return (name && process.env[name]) || ''; //
      },
    ),
  );
}

/**
 * コマンドラインを解析して必要な情報を取得する。
 *
 * @export
 * @param {Iterable<string>} [args]
 * @returns
 */
export function parseOptions(args?: Iterable<string>) {
  // コマンドライン解析
  const options = parse(
    {
      help: {
        alias: 'h',
        type: 'boolean',
        alone: true,
        describe: 'このヘルプを表示します。',
      },
      port: {
        alias: 'p',
        type: 'number',
        default: 50000,
        example: '#listen_port',
        describe: 'サーバーのポート番号を指定します。',
      },
      index: {
        alias: 'i',
        type: 'string',
        default: 'index.html/index.htm',
        example: 'index.html',
        describe: `
      URLが/で終わっていた場合に補完されるファイル名を指定します。
      /で区切ることで複数のファイル名を指定できます。
    `,
      },
      [unnamed]: {
        example: '/location/to/resource=path/to/file/or/directory',
        describe: `
      レスポンスとして返すファイルの位置を設定するエントリーを指定します。
      {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
      URLのパスは/から始まっている必要があります。
      ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
      ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
      ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
      ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
      ファイルは指定された順に検索していき、最初に見つかったものを返します。
    `,
      },
      [helpString]: {
        showUsageOnError: true,
        describe: `
    簡易HTTPサーバーを起動します。
    `,
      },
    },
    args,
  );

  // --help/-hが指定されたときはヘルプ文字列を出力して終了
  if ('help' in options) {
    console.log(options[helpString]);
    process.exit(0);
  }

  // ファイルエントリの解析
  const errors: string[] = [];
  const files = options[unnamed]
    .map(entry => {
      const eq = entry.indexOf('=');
      if (eq < 0) {
        // =がない
        errors.push(`エントリーに\`='がありません。: ${entry}`);
        return undefined;
      }
      if (eq === 0) {
        // locationがない
        errors.push(`locationが空です。: ${entry}`);
        return undefined;
      }
      if (eq === entry.length - 1) {
        // pathがない
        errors.push(`pathが空です。: ${entry}`);
        return undefined;
      }
      const location = entry.slice(0, eq);
      if (location.charAt(0) !== '/') {
        // pathがない
        errors.push(`locationは\`/'で始まっている必要があります。: ${entry}`);
        return undefined;
      }
      const path = resolvePathSpec(entry.slice(eq + 1));
      if (!existsSync(path)) {
        errors.push(`pathが見つかりません。: ${entry}`);
        return undefined;
      }
      return [location, path] as [string, string];
    })
    // undefinedを除外
    .filter(<T>(o: T): o is NonNullable<T> => !!o);
  // エラーがなくてもエントリーがなければエラー
  if (!errors.length && !files.length) {
    errors.push(`エントリーが指定されていません。`);
  }

  // エラーがあったら使い方を表示して終了
  if (errors.length) {
    console.error(errors.map(s => `${s}\n`).join(''));
    console.log(options[helpString]);
    process.exit(-1);
  }
  return {
    port: options.port,
    index: options.index.split('/'),
    files,
  };
}
