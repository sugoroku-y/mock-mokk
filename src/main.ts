import {mokkServer} from './mokkServer';
import {parseOptions} from './parseOptions';

const {port, index, files} = parseOptions();
void mokkServer(port, index, files);
