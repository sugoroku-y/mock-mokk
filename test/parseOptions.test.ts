import {homedir} from 'os';
import {resolve} from 'path';
import {parseOptions} from '../src/parseOptions';
import './toExitProcess';
import './toConsoleOutputMatchingSnapshot';

describe('parseOptions', () => {
  test('minimum', () => {
    expect(parseOptions(['/=.'])).toEqual({
      port: 50000,
      index: ['index.html', 'index.htm'],
      files: [['/', resolve('.')]],
    });
  });
  test('homedir', () => {
    expect(parseOptions(['/=$/'])).toEqual({
      port: 50000,
      index: ['index.html', 'index.htm'],
      files: [['/', homedir()]],
    });
  });
  test('dollar', () => {
    expect(parseOptions(['/=test/$$'])).toEqual({
      port: 50000,
      index: ['index.html', 'index.htm'],
      files: [['/', resolve('test', '$')]],
    });
  });
  test('env', () => {
    process.env['TESTENV'] = 'test/$';
    expect(parseOptions(['/=$TESTENV'])).toEqual({
      port: 50000,
      index: ['index.html', 'index.htm'],
      files: [['/', resolve('test', '$')]],
    });
    expect(parseOptions(['/=${TESTENV}'])).toEqual({
      port: 50000,
      index: ['index.html', 'index.htm'],
      files: [['/', resolve('test', '$')]],
    });
    delete process.env['TESTENV'];
    expect(parseOptions(['/=test/$TESTENV$'])).toEqual({
      port: 50000,
      index: ['index.html', 'index.htm'],
      files: [['/', resolve('test', '$')]],
    });
  });
  test('help', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions(['--help'])).toExitProcess(0);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
  test('no entry', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions([])).toExitProcess(-1);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
  test('no equal', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions(['test'])).toExitProcess(-1);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
  test('empty location', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions(['=test'])).toExitProcess(-1);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
  test('empty path', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions(['test='])).toExitProcess(-1);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
  test('not start slash', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions(['test=test'])).toExitProcess(-1);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
  test('not exist', () => {
    expect(() => {
      expect(() => {
        expect(() => parseOptions(['/test=test/not-exist'])).toExitProcess(-1);
      }).toConsoleOutputMatchingSnapshot('log');
    }).toConsoleOutputMatchingSnapshot('error');
  });
});
