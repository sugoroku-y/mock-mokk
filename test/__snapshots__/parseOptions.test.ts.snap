// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseOptions empty location: toConsoleOutputMatchingSnapshot:error 1`] = `
Array [
  Array [
    "locationが空です。: =test
",
  ],
]
`;

exports[`parseOptions empty location: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;

exports[`parseOptions empty path: toConsoleOutputMatchingSnapshot:error 1`] = `
Array [
  Array [
    "pathが空です。: test=
",
  ],
]
`;

exports[`parseOptions empty path: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;

exports[`parseOptions help: toConsoleOutputMatchingSnapshot:error 1`] = `Array []`;

exports[`parseOptions help: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;

exports[`parseOptions no entry: toConsoleOutputMatchingSnapshot:error 1`] = `
Array [
  Array [
    "エントリーが指定されていません。
",
  ],
]
`;

exports[`parseOptions no entry: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;

exports[`parseOptions no equal: toConsoleOutputMatchingSnapshot:error 1`] = `
Array [
  Array [
    "エントリーに\`='がありません。: test
",
  ],
]
`;

exports[`parseOptions no equal: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;

exports[`parseOptions not exist: toConsoleOutputMatchingSnapshot:error 1`] = `
Array [
  Array [
    "pathが見つかりません。: /test=test/not-exist
",
  ],
]
`;

exports[`parseOptions not exist: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;

exports[`parseOptions not start slash: toConsoleOutputMatchingSnapshot:error 1`] = `
Array [
  Array [
    "locationは\`/'で始まっている必要があります。: test=test
",
  ],
]
`;

exports[`parseOptions not start slash: toConsoleOutputMatchingSnapshot:log 1`] = `
Array [
  Array [
    "Version: mock-server 0.0.1
Usage:
  npx mock-server [--port #listen_port] [--index index.html] [--] [/location/to/resource=path/to/file/or/directory...]
  npx mock-server --help

Description:
  簡易HTTPサーバーを起動します。

Options:
  --help, -h
    このヘルプを表示します。
  --port, -p #listen_port
    サーバーのポート番号を指定します。
  --index, -i index.html
    URLが/で終わっていた場合に補完されるファイル名を指定します。
    /で区切ることで複数のファイル名を指定できます。
  [--] [/location/to/resource=path/to/file/or/directory...]
    レスポンスとして返すファイルの位置を設定するエントリーを指定します。
    {URLのパス部分}={割り当てられるファイル、もしくはディレクトリーの位置}の形式で指定します。
    URLのパスは/から始まっている必要があります。
    ファイル/ディレクトリーに相対パスを指定した場合は実行時のカレントディレクトリを基準とします。
    ファイル/ディレクトリーに$/で始まるパスを指定した場合はホームディレクトリに$/以降のパスを連結したものになります。
    ファイル/ディレクトリーには$ENVの形で環境変数を指定することができます。
    ファイル/ディレクトリーに$そのものを指定したい場合には$$としてください。
    ファイルは指定された順に検索していき、最初に見つかったものを返します。
",
  ],
]
`;
